src/main/flight/autopilot.c

```diff
@@ -91,17 +206,182 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+        ap.gpsDataFreqHz = getGpsDataFrequencyHz();
+
+        // first get NS and EW distances from current location (gpsSol.llh) to target location
+        vector2_t gpsDistance;
+        GPS_latLongVectors(&gpsSol.llh, &currentTargetLocation, &gpsDistance.x, &gpsDistance.y); // X is EW, Y is NS
+        ap.efAxis[EW].distance = gpsDistance.x;
+        ap.efAxis[NS].distance = gpsDistance.y;
+
+        const float distanceCm = vector2Norm(&gpsDistance);
+
+        const float leak = 1.0f - 0.4f * ap.gpsDataIntervalS;
+        // leak iTerm while sticks are centered, 2s time constant approximately
+        const float lpfGain = pt1FilterGain(ap.pt1Cutoff, ap.gpsDataIntervalS);
+
+        // ** Sanity check **
+        // primarily to detect flyaway from no Mag or badly oriented Mag
+        // must accept some overshoot at the start, especially if entering at high speed
+        if (distanceCm > ap.sanityCheckDistance) {
+            return false;
+        }
+
+        static float prevPidDASquared = 0.0f; // if we limit DA on true vector length
+        const float maxDAAngle = 35.0f; // limit in degrees; arbitrary angle
+        const float sqMaxDAAngle = sq(maxDAAngle);
+
+        for (axisEF_t loopAxis = EW; loopAxis <= NS; loopAxis++) {
+            earthFrame_t *efAxis = &ap.efAxis[loopAxis];
+            // separate PID controllers for latitude (NorthSouth or NS) and longitude (EastWest or EW)
+
+            // ** P **
+            const float pidP = efAxis->distance * positionPidCoeffs.Kp;
+
+            // ** I **
+            efAxis->integral += efAxis->isStopping ? 0.0f : efAxis->distance * ap.gpsDataIntervalS;
+            // only add to iTerm while in hold phase
+            const float pidI = efAxis->integral * positionPidCoeffs.Ki;
+
+            // ** D ** //
+            // Velocity derived from GPS position works better than module supplied GPS Speed and Heading information
+
+            float velocity = (efAxis->distance - efAxis->previousDistance) * ap.gpsDataFreqHz; // cm/s, minimum step 11.1 cm/s
+            efAxis->previousDistance = efAxis->distance;
+            pt1FilterUpdateCutoff(&efAxis->velocityLpf, lpfGain);
+            const float velocityFiltered = pt1FilterApply(&efAxis->velocityLpf, velocity);
+            float pidD = velocityFiltered * positionPidCoeffs.Kd;
+
+            float acceleration = (velocity - efAxis->previousVelocity) * ap.gpsDataFreqHz;
+            efAxis->previousVelocity = velocity;
+            pt1FilterUpdateCutoff(&efAxis->accelerationLpf, lpfGain);
+            const float accelerationFiltered = pt1FilterApply(&efAxis->accelerationLpf, acceleration);
+            const float pidA = accelerationFiltered * positionPidCoeffs.Kf;
+
+            if (ap.sticksActive) {
+                // sticks active 'phase'
+                efAxis->isStopping = true;
+                updateLocation(efAxis, loopAxis);
```

Comment:
The code is this way because while the sticks are active, things have to be done per-axis, eg leaking the iTerm, and resetting the location each time the aircraft changes it's position, to avoid large D step when the sticks stop moving.

It also simplifies the logic for the 'stopping' phase which is mutually exclusive of the active stick movement phase.

I tried several alternatives and this seemed most compact - at least with my limited programming skill.

Please propose exactly how to better write this part of the code.